flink-core/src/main/java/org/apache/flink/core/plugin/DirectoryBasedPluginFinder.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.plugin;
+
+import org.apache.flink.util.function.FunctionUtils;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to create a collection of {@link PluginDescriptor} based on directory structure for a given plugin
+ * root folder.
+ *
+ * <p>The expected structure is as follows: the given plugins root folder, containing the plugins folder. One plugin folder
+ * contains all resources (jar files) belonging to a plugin. The name of the plugin folder becomes the plugin id.
+ * <pre>
+ * plugins-root-folder/
+ *	         |------------plugin-a/ (folder of plugin a)
+ *	         |	              |-plugin-a-1.jar (the jars containing the classes of plugin a)
+ *	         |                |-plugin-a-2.jar
+ *	         |                |-...
+ *	         |
+ *	         |------------plugin-b/
+ *		     |                |-plugin-b-1.jar
+ *		    ...               |-...
+ * </pre>
+ */
+public class DirectoryBasedPluginFinder implements PluginFinder {
+
+	private static final PathMatcher JAR_FILE_MATCHER =
+		FileSystems.getDefault().getPathMatcher("glob:**.jar");
+
+	private final Path pluginsRootDir;
+
+	public DirectoryBasedPluginFinder(Path pluginsRootDir) {
+		this.pluginsRootDir = pluginsRootDir;
+	}
+
+	@Override
+	public Collection<PluginDescriptor> findPlugins() throws IOException {
+
+		if (!Files.isDirectory(pluginsRootDir)) {
+			throw new IOException("Plugin directory " + pluginsRootDir + " does not exist!");
```

Comment:
nitty nit: I would also add some brackets around the paths here, like `"Plugin directory [" + pluginsRootDir + "] does not exist!"`. It saves some confusion for the users in some edge cases (like white space names, empty or some directory name that blends into the exception message like `probably` ;) )

same in L93